### ISSUE TITLE
Upgrade and update instructions

### ISSUE THREAD

**Issue by SystemParadox** [ideal-postcodes/postcodes.io#63]
Hi,

Thanks very much for this project, been using it for about a year now and everything is great.

However, one of our users has reported some missing postcodes so I went to see about finally updating the postcode database...

First issue, there are no update instructions, only creating a new database from scratch. Updating using the setup script doesn't work because it throws errors when the database already exists.

I noticed the latest master has a `bin/update_ons.js`, so I went and updated to the latest master. 

Unfortunately, this now gives `error: relation "districts" does not exist`. So clearly the database structure has changed a bit. Fair enough, but how do I fix this? Is there a migration script somewhere or do I have to delete the whole database and start from scratch?

Please could we get some clear instructions on how to update the postcodes data, and how to upgrade the server. Is it possible to safely upgrade a live server somehow? Or is the only option to drop everything and rebuild from scratch?

Thanks.

**Comment by metase** [comment]
I'm in a similar position, although I restore this into SQL server (crunch the dump file into CSV style and then import via BI tools) and then run a comparison. It's not ideal but does the trick (mostly)

**Comment by cblanc** [comment]
Hi there,

You stumbled onto the right script which updates the database. At least, this is the one we use internally for 0 downtime updates to postcodes.io. All it does is create and populate parallel relations before swapping it in for the old data by renaming the relations. However this uses the raw ONSPD CSV file and not the pg_dump we host on S3.

The error you brought up is definitely new, I'll look into this as well as rewriting the script with a view to documenting a clear path for updating the database. Just to be clear, is the error is being raised during the migration? Or is it being raised by the app?

This will take some time however. The alternative would be to (as you mentioned) delete and load from scratch which will result in ~10 minutes of downtime. If you know how to use pg_dump, you could feasibly update the data without downtime by piping the contents into a clean database and then redirecting the app's datasource when the update is complete.

**Comment by SystemParadox** [comment]
The error was raised by the app after I updated the code, but didn't change the database. Are you suggesting that the database structure hasn't changed at all? I was updating from over a year ago.

I ended up going with the dump option and the 10 minutes downtime. It wasn't too bad, but it would definitely be really nice to have some way to live update the existing database without any downtime.

Thanks.

**Comment by cblanc** [comment]
Yes the schema has changed. We've normalised some data to accommodate some feature requests which would have caused later iterations of the app to be incompatible with the old schema.

I'll be adding upgrade notes to README.md in case of any breaking changes (like the schema) and resume incrementing this repo's semver version number.

I think 0 downtime schema changes will be hard to pull off on the same database, since you'd have to coordinate upgrading and rebooting the server with the migration. Fyi, in our case we provisioned new servers and update DNS records, as our infrastructure makes this quite easy.

0 downtime data updates should be feasible however, I'll see to it that the ONS update script is documented
